core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinRemoveRule.java

```diff
@@ -78,47 +99,62 @@ public ProjectJoinRemoveRule(
     final Project project = call.rel(0);
     final Join join = call.rel(1);
     final boolean isLeftJoin = join.getJoinType() == JoinRelType.LEFT;
-    int lower = isLeftJoin
-        ? join.getLeft().getRowType().getFieldCount() : 0;
-    int upper = isLeftJoin
-        ? join.getRowType().getFieldCount()
-        : join.getLeft().getRowType().getFieldCount();
-
-    // Check whether the project uses columns whose index is between
-    // lower(included) and upper(excluded).
-    for (RexNode expr: project.getProjects()) {
-      if (RelOptUtil.InputFinder.bits(expr).asList().stream().anyMatch(
-          i -> i >= lower && i < upper)) {
-        return;
-      }
-    }
+    final boolean isRightJoin = join.getJoinType() == JoinRelType.RIGHT;
+    final boolean isInnerJoin = join.getJoinType() == JoinRelType.INNER;
 
-    final List<Integer> leftKeys = new ArrayList<>();
-    final List<Integer> rightKeys = new ArrayList<>();
-    RelOptUtil.splitJoinCondition(join.getLeft(), join.getRight(),
-        join.getCondition(), leftKeys, rightKeys,
-        new ArrayList<>());
+    // Check project range
+    ImmutableBitSet projectBits = RelOptUtil.InputFinder.bits(project.getProjects(), null);
+    final int leftFieldsNum = join.getLeft().getRowType().getFieldCount();
+    final boolean onlyUseLeft = projectBits.asList().stream()
+        .allMatch(i -> i >= 0 && i < leftFieldsNum);
+    final boolean onlyUseRight = projectBits.asList().stream().allMatch(i -> i >= leftFieldsNum
+            && i < join.getRowType().getFieldCount());
```

Comment:
To match the style you used above:
```suggestion
    final boolean onlyUseRight = projectBits.asList().stream()
        .allMatch(i -> i >= leftFieldsNum && i < join.getRowType().getFieldCount());
```